### PR TITLE
[Merged by Bors] - feat: universe generalizations in UniqueGluing and Forget

### PR DESCRIPTION
**Pull request by alreadydone**
+ Add a criterion `isLimit_iff` for a cone to be a limit in Limits/Types.lean.

+ Use the criterion to show the equivalence between the UniqueGluing and PairwiseIntersection sheaf conditions without going through EqualizerProducts, thereby generalize the universes.

+ Remove theorems/def that are now unnecessary. (cc @justus-springer)

+ Generalize `isSheaf_iff_isSheaf_forget` for sheaves on sites using `isSheaf_iff_isLimit`; use it to prove and generalize the result on topological spaces, removing a large chunk of code.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by alreadydone** [comment]
@kbuzzard I've been able to generalize the universes in `isSheaf_iff_isSheaf_forget` which you requested in the comments in [#5659](https://github.com/leanprover-community/mathlib4/pull/5659) in https://github.com/leanprover-community/mathlib4/pull/7654/commits/05b5bb61a29af22d6a10f177570ce3112278d6b0, using the sheaf condition `isSheaf_iff_isLimit` I introduced [two years ago](https://github.com/leanprover-community/mathlib/commit/343cbd9867860506f5daded9173cc1a3bc064906). This sheaf condition in terms of limits is very convenient as it doesn't require any extra conditions (pullbacks in C, products in A), and is manifestly equivalent to the OpensLeCover condition on topological spaces; I should probably add it to the module docstring to make it more well-known. When C has pullbacks, we could use the Pairwise diagram to define a "PairwisePullback" sheaf condition, which would have the advantage of not requiring [HasProducts A], which the equalizer sheaf condition requires. Since the category of opens has pullbacks, the [PairwiseIntersections](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Sheaves/SheafCondition/PairwiseIntersections.html#TopCat.Presheaf.IsSheafPairwiseIntersections) condition is also unconditionally equivalent to the usual sheaf condition, and in this PR I have successfully used it to generalize universes in the UniqueGluing sheaf condition.

**Comment by alreadydone** [comment]
Pinging @semorrison for review as this finishes the incomplete universe generalization in #5726.

**Comment by semorrison** [comment]
> Pinging @semorrison for review as this finishes the incomplete universe generalization in #5726.

Thanks, but I don't have time to review this at the moment.

File: Mathlib/CategoryTheory/Limits/Types.lean [hunk 1]
```diff
@@ -31,6 +31,58 @@ universe v u
 
 namespace CategoryTheory.Limits.Types
 
+section limit_characterization
+
+variable {J : Type v} [Category J] {F : J ⥤ Type u}
+
+/-- Given a section of a functor F into `Type*`,
+  construct a cone over F with `PUnit` as the cone point. -/
+def cone_of_section {s} (hs : s ∈ F.sections) : Cone F where
```
Comment (erdOne) [inline Mathlib/CategoryTheory/Limits/Types.lean:40]: I'm not familiar with the new naming conventions at all, but should this be `coneOfSection`?

File: Mathlib/CategoryTheory/Limits/Types.lean [hunk 2]
```diff
@@ -31,6 +31,58 @@ universe v u
...
+  pt := PUnit
+  π :=
+  { app := fun j _ ↦ s j,
+    naturality := fun i j f ↦ by ext; exact (hs f).symm }
+
+/-- Given a cone over a functor F into `Type*` and an element in the cone point,
+  construct a section of F. -/
+def section_of_cone (c : Cone F) (x : c.pt) : F.sections :=
```
Comment (erdOne) [inline Mathlib/CategoryTheory/Limits/Types.lean:48]: Maybe `sectionOfCone`?

File: Mathlib/Topology/Sheaves/SheafCondition/UniqueGluing.lean [hunk 3]
```diff
@@ -89,122 +88,66 @@ end
 
 section TypeValued
 
-variable {X : TopCat.{x}} (F : Presheaf (Type u) X) {ι : Type x} (U : ι → Opens X) [UnivLE.{x, u}]
+variable {X : TopCat.{x}} {F : Presheaf (Type u) X} {ι : Type x} {U : ι → Opens X}
+
+/-- Given sections over a family of open sets, extend it to include
+  sections over pairwise intersections of the open sets. -/
+def obj_pairwise_of_family (sf : ∀ i, F.obj (op (U i))) :
```
Comment (erdOne) [inline Mathlib/Topology/Sheaves/SheafCondition/UniqueGluing.lean:95]: `objPairwiseOfFamily`?

**Review by erdOne — APPROVED** [review]
LGTM modulo the comments that I'm not sure about at all.

**Comment by erdOne** [comment]
> I think the equalizer condition (and EqualizerProducts for topological spaces) could be superseded by such a PairwisePullback condition and deprecated.

The equalizer condition is quite a prominent one in the literature and I think it should be kept nonetheless?

**Comment by alreadydone** [comment]
> The equalizer condition is quite a prominent one in the literature and I think it should be kept nonetheless?

I think it could go into Archive/ because I just don't see how it could be useful for mathlib development. It's basically just a version of PairwiseIntersection that requires an extra condition (existence of products), and I don't see any advantage working with equalizer diagram between products vs. working directly with the pairwise diagram. Rather than taking the products and then equalizer, it's easier to take the limit of a diagram in one stretch, and this is the approach used to do sheafification (where [WalkingMultispan](https://leanprover-community.github.io/mathlib4_docs/Mathlib/CategoryTheory/Limits/Shapes/Multiequalizer.html#CategoryTheory.Limits.WalkingMultispan.instInhabitedWalkingMultispan) is a generalization of the Pairwise category). (The only exception is probably when all intersections are empty, so the limit is actually a product.)

The existence of the product-equalizer condition in mathlib seems to lead to confusion and less general results with unnecessary HasProduct assumptions. Probably due to its centrality in the literature people tends to develop APIs around it and make it a hub to connect sheaf conditions, which led to less general results. I think there should at least be a warning that its use is discouraged, and PairwiseIntersections/Pullbacks should be used instead.

File: Mathlib/CategoryTheory/Limits/Types.lean (hunk 1 above)
Comment (alreadydone) [reply Mathlib/CategoryTheory/Limits/Types.lean]: Thanks, renamed

File: Mathlib/Topology/Sheaves/SheafCondition/UniqueGluing.lean [hunk 4]
```diff
@@ -89,123 +88,66 @@ end
...
+/-- Given sections over a family of open sets, extend it to include
+  sections over pairwise intersections of the open sets. -/
+def objPairwiseOfFamily (sf : ∀ i, F.obj (op (U i))) :
+    ∀ i, ((Pairwise.diagram U).op ⋙ F).obj i
+  | ⟨Pairwise.single i⟩ => sf i
+  | ⟨Pairwise.pair i j⟩ => F.map (infLELeft (U i) (U j)).op (sf i)
+
+/-- Given a compatible family of sections over open sets, extend it to a
+  section of the functor `(Pairwise.diagram U).op ⋙ F`. -/
+def IsCompatible.section_pairwise {sf} (h : IsCompatible F U sf) :
```
Comment (erdOne) [inline Mathlib/Topology/Sheaves/SheafCondition/UniqueGluing.lean:102]:
```suggestion
def IsCompatible.sectionPairwise {sf} (h : IsCompatible F U sf) :
```

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**